Generate complete bindings to libzmq

JamesWrigley opened the pull request:
This is a rather invasive change to use Clang.jl to generate bindings for everything in `zmq.h`, and then use those bindings in the rest of the library. My motivation for doing this is to make it easier to call the libzmq functions, both for us and for anyone who wants to use the low-level functions (e.g. for things we haven't explicitly wrapped). It's also nice to have the constants generated for us.

It touches a lot of internals so this would definitely benefit from a review from someone more familiar with them :sweat_smile: In particular the code for `_Message` and `Message`, which took me a while to understand. I've written down my understanding of how that works in the comments of `gen.jl`. For the reviewers: I've tried to keep the commits atomic, so it might be easiest to review the commits one-by-one. ~~I don't think it's quite ready to merge yet because I'd like to add some more docs, but it should be ready for review.~~

CC @stevengj, @vtjnash

stevengj commented:
Overall, I'm in favor of this, though I haven't had a chance to review it in detail.

JamesWrigley commented:
I made another change to the internals in 56d3a9b to force using `Ref{_Message}` everywhere. I *think* this is correct because it would be unsafe to use most/all of those functions with a plain `_Message` since it's a immutable struct and doesn't need to have a stable address.

JamesWrigley commented:
Added docs in 6729eba, and fixed the use of a deprecated function in ef30b30. Pending review, I think this is ready to merge :crossed_fingers: 

(P.S., please don't merge it with the fixup commits, I'll rebase beforehand)

JamesWrigley commented:
Rebased the branch, and modified the generated bindings to use `ccall()` instead of `@ccall` in c13987d (for compatibility with Julia 1.3).

JamesWrigley commented:
(bump)
